File: Python3/1 - 200/200 Number of Islands.py
# 132ms, 99.35%; 15.1MB, 20.60%
class Solution:
    def numIslands(self, grid: list) -> int:
        res = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] != '1': continue
                res += 1
                self.visit(grid, i, j)
        return res

    def visit(self, grid: list, i: int, j: int) -> None:
        grid[i][j] = '2'
        if i > 0 and grid[i - 1][j] == '1': self.visit(grid, i - 1, j)
        if j > 0 and grid[i][j - 1] == '1': self.visit(grid, i, j - 1)
        if i + 1 < len(grid) and grid[i + 1][j] == '1': self.visit(grid, i + 1, j)
        if j + 1 < len(grid[0]) and grid[i][j + 1] == '1': self.visit(grid, i, j + 1)
        

File: 1 - 200/200 Number of Islands.cc
// 4ms, 99.94%; 11MB, 44.69%
auto x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution
{
public:
    int numIslands(vector<vector<char>> &grid)
    {
        int res = 0;
        for (int i = 0, row = grid.size(); i < row; ++i) {
            for (int j = 0, col = grid[0].size(); j < col; ++j) {
                if (grid[i][j] != '1') continue;
                ++res;
                visit(grid, i, j);
            }
        }
        return res;
    }

private:
    void visit(vector<vector<char>> &grid, int i, int j)
    {
        grid[i][j] = '2';
        if (i > 0 && grid[i - 1][j] == '1') visit(grid, i - 1, j);
        if (j > 0 && grid[i][j - 1] == '1') visit(grid, i, j - 1);
        if (i + 1 < grid.size() && grid[i + 1][j] == '1') visit(grid, i + 1, j);
        if (j + 1 < grid[0].size() && grid[i][j + 1] == '1') visit(grid, i, j + 1);
    }
};
